server/src/test/java/org/apache/druid/server/scheduling/HiLoQueryLaningStrategyTest.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.query.Druids;
+import org.apache.druid.query.QueryContexts;
+import org.apache.druid.query.QueryPlus;
+import org.apache.druid.query.aggregation.CountAggregatorFactory;
+import org.apache.druid.query.timeseries.TimeseriesQuery;
+import org.apache.druid.server.QueryLaningStrategy;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class HiLoQueryLaningStrategyTest
+{
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  private Druids.TimeseriesQueryBuilder queryBuilder;
+  private HiLoQueryLaningStrategy strategy;
+
+  @Before
+  public void setup()
+  {
+    this.queryBuilder = Druids.newTimeseriesQueryBuilder()
+                              .dataSource("test")
+                              .intervals(ImmutableList.of(Intervals.ETERNITY))
+                              .granularity(Granularities.DAY)
+                              .aggregators(new CountAggregatorFactory("count"));
+
+    this.strategy = new HiLoQueryLaningStrategy(10);
+  }
+
+  @Test
+  public void testMaxLowThreadsRequired()
+  {
+    expectedException.expect(NullPointerException.class);
+    expectedException.expectMessage("maxLowThreads must be set");
+    QueryLaningStrategy strategy = new HiLoQueryLaningStrategy(null);
+  }
+
+  @Test
+  public void testLaneLimits()
+  {
+    Object2IntMap<String> laneConfig = strategy.getLaneLimits();
+    Assert.assertEquals(1, laneConfig.size());
+    Assert.assertTrue(laneConfig.containsKey(HiLoQueryLaningStrategy.LOW));
+    Assert.assertEquals(10, laneConfig.getInt(HiLoQueryLaningStrategy.LOW));
+  }
+
+  @Test
+  public void testLaningNoPriority()
+  {
+    TimeseriesQuery query = queryBuilder.build();
+    Assert.assertFalse(strategy.computeLane(QueryPlus.wrap(query), ImmutableSet.of()).isPresent());
+  }
+
+  @Test
+  public void testLaningZeroPriority()
+  {
+    TimeseriesQuery query = queryBuilder.context(ImmutableMap.of(QueryContexts.PRIORITY_KEY, 0)).build();
+    Assert.assertFalse(strategy.computeLane(QueryPlus.wrap(query), ImmutableSet.of()).isPresent());
+  }
+
+  @Test
+  public void testLaningInteractivePriority()
+  {
+    TimeseriesQuery query = queryBuilder.context(ImmutableMap.of(QueryContexts.PRIORITY_KEY, 100)).build();
+    Assert.assertFalse(strategy.computeLane(QueryPlus.wrap(query), ImmutableSet.of()).isPresent());
+  }
+
+  @Test
+  public void testLaningLowPriority()
+  {
+    TimeseriesQuery query = queryBuilder.context(ImmutableMap.of(QueryContexts.PRIORITY_KEY, -1)).build();
+    Assert.assertTrue(strategy.computeLane(QueryPlus.wrap(query), ImmutableSet.of()).isPresent());
+    Assert.assertEquals(
+        HiLoQueryLaningStrategy.LOW,
+        strategy.computeLane(QueryPlus.wrap(query), ImmutableSet.of()).get()
+    );
+  }
+}
```

Comment:
Hmm, I'm not yet quite sure if not allowing a laning strategy to override the manually set context value is the correct behavior yet, so I didn't add a test. 

Do you have any thoughts on whether or not to replace an explicit, user specified lane be delegated to the laning strategy? The more I think about it the more I think the laning strategy should get to choose, so I think I'm going to change the behavior. If we still think that hilo should honor the user specified lane, then I can add a test for that (or add a test that it is overridden if we decide to go in the other direction).

Reply:
Since the user can select a lane by setting the priority in the query context, I'm not sure of the benefit of also having a lane in the query context. Later, when there's automatic setting of the lane/priority, I think it should first respect the value in the query context if there's one before applying the automatic behavior.

Reply:
I think there are a few reasons to put it on the query context. For one to propagate the value downstream to historical and realtime tasks so they can utilize that information in the future to also make load management decisions. 

But also, I'm not sure that lane necessarily _has_ to be related to priority just because the current implementation I have provided is. It is really just a label used to enforce limits. 

In my list of potential follow-up work, I wanted to support a manual laning strategy that only supports explicit use defined lanes on the context [both to make integration tests easier](https://github.com/clintropolis/druid/compare/query-laning-and-load-shedding...clintropolis:manual-query-laning) but also just to account for scenarios where an external application drives these decisions and still wishes to enforce limits on classes of queries. I also suggested a tier based laning strategy, which would not be related to priority, but rather the set of servers it is going to be querying.